### ISSUE TITLE
lambdify passes the wrong arguments to numpy.log

### ISSUE THREAD

**Issue by simonLeary42** [sympy/sympy#25778]
```py
import numpy as np
import sympy as sp
from sympy.parsing.latex import parse_latex

expr = parse_latex(r"\ln(x)")
func = sp.lambdify('x', expr, 'numpy')
x_vals = np.linspace(0.1, 10, 400)
y_vals = func(x_vals)
```

```
TypeError: return arrays must be of ArrayType
```

I think it should be doing `np.math.log` where it is currently doing `np.log`.

**Comment by simonLeary42** [comment]
https://github.com/sympy/sympy/blob/368eb0f483f2b4cd84dc0de3edc7abdc876d3d46/sympy/parsing/latex/_parse_latex_antlr.py#L446

**Comment by simonLeary42** [comment]
This also produces the same behavior:
```py
import numpy as np
import sympy as sp
expr = sp.log('x', sp.E, evaluate=False)
func = sp.lambdify('x', expr, 'numpy')
func(10)
```

**Comment by simonLeary42** [comment]
Here is a hacky workaround: https://github.com/sympy/sympy/commit/bfffc90827fce71281d359bdd6fef4962abe1b5b

**Comment by simonLeary42** [comment]
This:
```py
MODULES["numpy"][2] += "; from numpy import lib.scimath.logn"
NUMPY_TRANSLATIONS["log"] = "logn"
```
was almost a drop in solution.

but the problem is that `numpy.lib.scimath.logn` expects its arguments to be in "base, input" order while sympy expects its arguments to be in "input, base" order.

**Comment by bjodah** [comment]
My guess is that `NumPyPrinter` needs a `_print_log(self, item)` method, which inspects `len(item.args)` and emits `logn` if `len == 2`, and the default `log(...)` call otherwise (when length is one).

**Comment by asmeurer** [comment]
sympy.log(x, b, evaluate=False) is currently not supported. See https://github.com/sympy/sympy/issues/6888

**Comment by asmeurer** [comment]
What is the plan for making lark the default? I'd like to remove the antlr parser entirely.

**Comment by wermos** [comment]
I have no problems with making the change. The Lark-based parser is on par with the ANTLR parser except for a couple of fringe features. On the other hand, it has quite a few improvements that the ANTLR-based parser lacks.

**Comment by oscarbenjamin** [comment]
> sympy.log(x, b, evaluate=False) is currently not supported. See #6888

It should be supported. This is a bug in lambdify and the proper fix should be as suggested by @bjodah above:

> `NumPyPrinter` needs a `_print_log(self, item)` method, which inspects `len(item.args)` and emits `logn` if `len == 2`, and the default `log(...)` call otherwise

**Comment by asmeurer** [comment]
See the discussion in the linked issue, particularly the most recent comment https://github.com/sympy/sympy/issues/6888#issuecomment-1696863426. We need to figure out the correct fix for this before we go changing things ad hoc. The solution suggested by @bjodah corresponds to one of the potential fixes (leaving `log` as both 1- and 2-arg), but I'm not yet sure if it's the right one.

**Comment by oscarbenjamin** [comment]
> See the discussion in the linked issue, particularly the most recent comment [#6888 (comment)](https://github.com/sympy/sympy/issues/6888#issuecomment-1696863426). We need to figure out the correct fix for this before we go changing things ad hoc.

I don't think that those considerations really affect whether or not `lambdify` should check for a 2-arg log and then handle it properly. In some way there should be a 2-arg log and however that is represented it should be handled by lambdify. If in future log is changed to two different classes or something then `lambdify` should be changed at the same time to ensure that it can pick up on whichever representations are being used for either 1- or 2-arg logs. In the meantime lambdify should be fixed to handle the current representation of a 2-arg log.

> The solution suggested by @bjodah corresponds to one of the potential fixes (leaving `log` as both 1- and 2-arg), but I'm not yet sure if it's the right one.

The suggestion by @bjodah is just to fix lambdify so that it works with the way that a 2-arg log is currently represented. To say that we should not fix that until deciding whether or not to change the representation of log is backwards. We should fix lambdify to work with the expressions as they currently exist in SymPy. Then if we want to change the representations of the expressions in future we should change lambdify at that point as well.

This is a good example of what I was referring to in https://github.com/sympy/sympy/pull/25790#issuecomment-1768211167:

> The problem as I see it is that when these issues come up the suggestion is usually that "broken" unevaluated expressions should not be used rather than to fix the actual problem.

**Comment by asmeurer** [comment]
lambdify is just one function. Every single part of SymPy that does something with `log` is wrong for 2-arg "unevaluated" logs. 

If you read my comment, there are basically two options to fix this:

- Fix the codebase to handle 2-arg logs (and simultaneously make 2-arg log not evaluate). 
- Add a separate class for 2-arg log, and make 2-arg `log` no longer a thing. 

The reason why we shouldn't *necessarily* "just do" the first option is because it's mutually exclusive with the second option, and the second might be preferable because it maintains backwards compatibility more.

So yes, these considerations do matter. We need to decide which one we want to do before we just jump in with one or the other.

**Comment by oscarbenjamin** [comment]
> If you read my comment, there are basically two options to fix this:
> 
> * Fix the codebase to handle 2-arg logs (and simultaneously make 2-arg log not evaluate).
> * Add a separate class for 2-arg log, and make 2-arg `log` no longer a thing.
> 
> The reason why we shouldn't _necessarily_ "just do" the first option is because it's mutually exclusive with the second option

Neither of those is what is being suggested here. The situation here is that someone can already do `log(a, b, evaluate=False)` but `lambdify` does not handle that. We can make `lambdify` work without doing either of the two things that you have described as being the two options. Fixing `lambdify` for this case currently would also not prevent anyone in future from doing either of the two things that you have suggested. There is no mutual exclusivity that prevents fixing `lambdify` right now.

**Comment by asmeurer** [comment]
Oscar, I really feel like you're not following what I'm saying, and we seem to be talking past one another. You say "neither of those are being suggested here" but my first bullet point is *exactly* what you are suggesting. 

I really encourage you to read through #6888 and the comment I linked to in particular. Fixing the 2-arg log problem has backwards compatibility implications and we need to decide what level of potential breakage we are OK with. I'll summarize here too:

1. Option 1: Make `log(x, b)` not evaluate. `log` is now a class that can have 1 or 2 args. 
2. Option 2: Make `log` (the class) always have 1 argument. There is a separate `log2` (or whatever name) class for $log_b(x)$. `log(x, b)` automatically creates `log2(x, b)`.

Option 1 is much simpler, because most code can just handle all logs uniformly with log base-e as just a special case. However, option 1 breaks any existing code that expects log to only be base-e (case in point, this exact issue). In particular, depending on how the code is written, the breakage could be a silent wrong result. 

Option 2 makes any existing code with these assumptions not return wrong results, because it sees `log2` as a completely separate class. But it's also arguably unnecessarily complicated because in reality 1-arg log is just a special case of 2-arg log. 

Right now, you can force the option 1 world by using evaluate=False, but no code actually expects that, so it's not really fair to say that it's already the case that we support it.

> There is no mutual exclusivity that prevents fixing lambdify right now.

There is, because if we go with the second option, then `log(x, b)` would create another class, `log2(x, b)`. So the fix to lambdify would do nothing, because it wouldn't expect that other class.

That's why we need to decide first if we want to go with option 1 or option 2.

Your "just fix lambdify" is effectively option 1. If decide we want to go with that, that's fine, but we need to be explicit about that, and it's not clear to me you've understood the implications of it.

**Comment by oscarbenjamin** [comment]
Regardless of whether `log(a, b)` evaluates its meaning is already defined in the codebase: the evaluation rule already defines its meaning. Therefore the meaning of `log(a, b, evaluate=False)` is also defined. That is already the situation now regardless of whether you consider the unevaluated `log(a, b)` to be "supported".

You are proposing to make one of two possible future changes either that `log(a, b)` no longer evaluates to `log(a)/log(b)` or that `log(a, b)` evaluates to a new expression type `log2(a, b)`. Neither of those potential future changes prevents fixing `lambdify` now to interpret `log(a, b)` in the way that the code base already interprets it.

> > here is no mutual exclusivity that prevents fixing lambdify right now.
> 
> There is, because if we go with the second option, then `log(x, b)` would create another class, `log2(x, b)`. So the fix to lambdify would do nothing, because it wouldn't expect that other class.

If we go with the second option at some point in future then `lambdify` can also be changed at *that* time to handle the new class. There is no reason that lambdify cannot be made to work with the expression types as they are *currently* implemented just because we might want to implement them differently in *future*. In a future change of the implementation of `log(a, b)` lambdify would need updating but we don't have to leave this bug sitting around indefinitely while contemplating those future changes.

(`log2` is a bad name for this hypothetical new expression type because it usually means log to the base 2 rather than a 2-arg log.)
